File: app/src/main/java/com/xephorium/crystalnote/ui/creation/CreationActivity.java
package com.xephorium.crystalnote.ui.creation;

import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.widget.EditText;
import android.widget.Toast;

import com.xephorium.crystalnote.R;
import com.xephorium.crystalnote.ui.base.BaseActivity;
import com.xephorium.crystalnote.ui.custom.NoteToolbar;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class CreationActivity extends BaseActivity implements CreationView {

    @BindView(R.id.creation_toolbar)
    NoteToolbar toolbar;

    CreationPresenter presenter;

//    @OnClick(R.id.creation_toolbar_save)
//    public void handleToolbarSaveClick() {
//        presenter.handleSaveClick(toolbarText.getText().toString(), contentText.getText().toString());
//    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.creation_activity_layout);
        ButterKnife.bind(this);

        presenter = new CreationPresenter(this);
        toolbar.setEditMode(true);
        toolbar.setLeftButtonImage(R.drawable.icon_back);
        toolbar.setRightButtonImage(R.drawable.icon_save);
        toolbar.setNoteToolbarListener(getCreationNoteToolbarListener());
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.attachView(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public Toolbar getToolbar() {
        return toolbar;
    }

    @Override
    public void onBackPressed() {
        presenter.handleBackClick();
    }

    @Override
    public void closeCreationActivity() {
        this.finish();
    }

    @Override
    public boolean isNewNotePopulated() {
        //return !TextUtils.isEmpty(toolbarText.getText()) || !TextUtils.isEmpty(contentText.getText());
        return false;
    }

    private NoteToolbar.NoteToolbarListener getCreationNoteToolbarListener() {
        return new NoteToolbar.NoteToolbarListener() {
            @Override
            public void onLeftButtonClick() {
                // presenter.handleBackClick(); // Calling Containing Class Methods From Here Causes UI Skipping?
                //Toast.makeText(getApplicationContext(), "Clicked", Toast.LENGTH_SHORT).show();
                //presenter.handleBackClick();
            }

            @Override
            public void onRightButtonClick() {

            }
        };
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/creation/CreationActivity.kt
package com.xephorium.crystalnote.ui.creation

import android.os.Bundle
import android.text.TextUtils

import com.xephorium.crystalnote.R
import com.xephorium.crystalnote.ui.base.ToolbarActivity
import com.xephorium.crystalnote.ui.custom.NoteToolbar

import kotlinx.android.synthetic.main.creation_activity_layout.*

class CreationActivity : ToolbarActivity(), CreationView {

    lateinit var presenter: CreationPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setActivityContent(R.layout.creation_activity_layout)

        presenter = CreationPresenter(this)
        setupCreationToolbar()
    }

    override fun onResume() {
        super.onResume()
        presenter.attachView(this)
    }

    public override fun onDestroy() {
        super.onDestroy()
        presenter.detachView()
    }

    override fun onBackPressed() {
        presenter.handleBackClick()
    }

    override fun closeCreationActivity() {
        this.finish()
    }

    override fun isNewNotePopulated(): Boolean {
        return creation_content.text?.let {
            !TextUtils.isEmpty(toolbar.titleContent) || !TextUtils.isEmpty(it)
        } ?: false
    }

    private fun setupCreationToolbar() {
        toolbar.isEditMode = true
        toolbar.setLeftButtonImage(R.drawable.icon_back)
        toolbar.setRightButtonImage(R.drawable.icon_save)
        toolbar.setNoteToolbarListener(getCreationNoteToolbarListener())
    }

    private fun getCreationNoteToolbarListener() = object : NoteToolbar.NoteToolbarListener {
        override fun onLeftButtonClick() {
            presenter.handleBackClick()
        }

        override fun onRightButtonClick() {
            creation_content.text?.let {
                presenter.handleSaveClick(toolbar.titleContent, it.toString())
            }
        }
    }
}
